In [ ]:
from google.colab import drive
drive.mount('/content/drive')


Mounted at /content/drive


In [ ]:
import pandas as pd
import matplotlib.pyplot as plt

In [ ]:
# Define file paths
us_city_data_path = '/content/drive/My Drive/Datathon/us_city_data.xlsx'
world_country_data_path = '/content/drive/My Drive/Datathon/world_country_data.xlsx'
nba_financial_data_path = '/content/drive/My Drive/Datathon/NBA_financial_data.xlsx'

# Read the Excel files into DataFrame objects
us_city_data = pd.read_excel(us_city_data_path)
world_country_data = pd.read_excel(world_country_data_path)
nba_financial_data = pd.read_excel(nba_financial_data_path)

# Display the first few rows of each DataFrame
print("US City Data:")
print(us_city_data.head())

print("\nWorld Country Data:")
print(world_country_data.head())

print("\nNBA Financial Data:")
print(nba_financial_data.head())


US City Data:
          city   city_ascii city_alt state_id  state_name  county_fips  \
0     New York     New York      NaN       NY    New York        36081   
1  Los Angeles  Los Angeles      NaN       CA  California         6037   
2      Chicago      Chicago      NaN       IL    Illinois        17031   
3        Miami        Miami      NaN       FL     Florida        12086   
4      Houston      Houston      NaN       TX       Texas        48201   

   county_name                county_fips_all  \
0       Queens  36081|36047|36085|36005|36061   
1  Los Angeles                           6037   
2         Cook                    17031|17043   
3   Miami-Dade                          12086   
4       Harris              48201|48157|48339   

                        county_name_all      lat  ...  poverty  \
0  Queens|Kings|Richmond|Bronx|New York  40.6943  ...     17.2   
1                           Los Angeles  34.1141  ...     16.6   
2                           Cook|DuPage  41.8375

In [ ]:
# from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
print("us_city_data: ")
print(us_city_data.columns)
print("world_country_data: ")
print(world_country_data.columns)
print("nba_financial_data: ")
print(nba_financial_data.columns)



us_city_data: 
Index(['city', 'city_ascii', 'city_alt', 'state_id', 'state_name',
       'county_fips', 'county_name', 'county_fips_all', 'county_name_all',
       'lat', 'lng', 'population', 'population_proper', 'density', 'source',
       'military', 'incorporated', 'cdp', 'township', 'timezone', 'ranking',
       'zips', 'id', 'age_median', 'age_under_10', 'age_10_to_19', 'age_20s',
       'age_30s', 'age_40s', 'age_50s', 'age_60s', 'age_70s', 'age_over_80',
       'age_over_65', 'age_18_to_24', 'age_over_18', 'male', 'female',
       'married', 'divorced', 'never_married', 'widowed', 'family_size',
       'family_dual_income', 'income_household_median',
       'income_household_under_5', 'income_household_5_to_10',
       'income_household_10_to_15', 'income_household_15_to_20',
       'income_household_20_to_25', 'income_household_25_to_35',
       'income_household_35_to_50', 'income_household_50_to_75',
       'income_household_75_to_100', 'income_household_100_to_150',
       '

In [ ]:
# Filter out cities without an NBA team in their state
cities_with_teams = nba_financial_data['State'].unique()
filtered_cities = us_city_data[~us_city_data['state_name'].isin(cities_with_teams)]
filtered_cities = filtered_cities[filtered_cities['state_name'] != 'District of Columbia']

# Sort cities based on population and income_household_median
sorted_cities = filtered_cities.sort_values(by=['population', 'income_household_median'], ascending=[False, False])

# Select the top 5 cities
top_cities_with_high_income = sorted_cities.head(5)

print("Top 5 cities without an NBA team based on population and income:")
print(top_cities_with_high_income[['city', 'state_name', 'population', 'income_household_median']].reset_index(drop=True))


Top 5 cities without an NBA team based on population and income:
        city   state_name  population  income_household_median
0    Seattle   Washington   3561397.0                 116068.0
1  Las Vegas       Nevada   2248047.0                  66356.0
2  Baltimore     Maryland   2196524.0                  58349.0
3  St. Louis     Missouri   2130881.0                  52941.0
4   San Juan  Puerto Rico   1814587.0                  26610.0


In [ ]:
import plotly.graph_objects as go

# Create population plot
population_trace = go.Bar(x=top_cities_with_high_income['city'], y=top_cities_with_high_income['population'], name='Population', marker_color='skyblue')
population_layout = go.Layout(title='Population of Top 5 Cities without NBA Teams', xaxis=dict(title='City'), yaxis=dict(title='Population'))

# Create income plot
income_trace = go.Bar(x=top_cities_with_high_income['city'], y=top_cities_with_high_income['income_household_median'], name='Median Household Income', marker_color='lightgreen')
income_layout = go.Layout(title='Median Household Income of Top 5 Cities without NBA Teams', xaxis=dict(title='City'), yaxis=dict(title='Median Household Income'))

# Show population plot
fig_population = go.Figure(data=[population_trace], layout=population_layout)
fig_population.update_xaxes(tickangle=45, tickfont=dict(size=10))

# Show income plot
fig_income = go.Figure(data=[income_trace], layout=income_layout)
fig_income.update_xaxes(tickangle=45, tickfont=dict(size=10))

# Show plots
fig_population.show()
fig_income.show()


### Let's now propose 2 new cities for NBA teams based on additional factors

In [ ]:
#'city',  'state_name', 'population',  'ranking','zips', 'age_median', 'male', 'female',
#'family_dual_income', 'income_household_median', 'income_individual_median', 'home_value', 'rent_median','unemployment_rate',
#'race_white', 'race_black', 'race_asian','race_native', 'race_pacific', 'race_other', 'race_multiple','hispanic',
#'commute_time', 'cbsa_fips', 'cbsa_name', 'cbsa_metro','csa_fips', 'csa_name'

In [ ]:
# Define weights for demographic features
weight_white = 0.2
weight_black = 0.2
weight_asian = 0.3
weight_native = 0.05
weight_pacific = 0.05
weight_other = 0.05
weight_multiple = 0.05
weight_hispanic = 0.3

# Combine race columns to create 'Total Demographics' with weighted sum
top_cities_with_high_income['Total Demographics'] = (
    top_cities_with_high_income['race_white'] * weight_white +
    top_cities_with_high_income['race_black'] * weight_black +
    top_cities_with_high_income['race_asian'] * weight_asian +
    top_cities_with_high_income['race_native'] * weight_native +
    top_cities_with_high_income['race_pacific'] * weight_pacific +
    top_cities_with_high_income['race_other'] * weight_other +
    top_cities_with_high_income['race_multiple'] * weight_multiple +
    top_cities_with_high_income['hispanic'] * weight_hispanic
)

# Combine economic indicators to create 'Economic Indicators'
top_cities_with_high_income['Economic Indicators'] = top_cities_with_high_income['income_household_median'] - top_cities_with_high_income['rent_median']

# Combine CBSA metro flag, commute time, and CSA information to create 'Geographic Factors'
# Define weights for each component
weight_cbsa_metro = 0.2
weight_commute_time = 0.1
weight_csa = 0.1

# Create 'Geographic Factors' column
top_cities_with_high_income['Geographic Factors'] = (top_cities_with_high_income['cbsa_metro'].astype(int) * weight_cbsa_metro) + (1 - top_cities_with_high_income['commute_time'] / top_cities_with_high_income['commute_time'].max()) * weight_commute_time + (top_cities_with_high_income['csa_fips'].notnull().astype(int) * weight_csa)

# Drop individual race columns since they are no longer needed
top_cities_with_high_income.drop(columns=['race_white', 'race_black', 'race_asian', 'race_native', 'race_pacific', 'race_other', 'race_multiple', 'hispanic'], inplace=True)

# Normalize the composite columns
from sklearn.preprocessing import MinMaxScaler

# Initialize the MinMaxScaler
scaler = MinMaxScaler()

# Normalize the composite columns
top_cities_with_high_income[['Total Demographics', 'Economic Indicators', 'Geographic Factors']] = scaler.fit_transform(top_cities_with_high_income[['Total Demographics', 'Economic Indicators', 'Geographic Factors']])

# weights
weight_demographics = 0.3
weight_economic = 0.4
weight_geographic = 0.3

# composite score for each city
top_cities_with_high_income['Composite Score'] = (top_cities_with_high_income['Total Demographics'] * weight_demographics) + \
                                                 (top_cities_with_high_income['Economic Indicators'] * weight_economic) + \
                                                 (top_cities_with_high_income['Geographic Factors'] * weight_geographic)

# Rank the cities and select the top 2
top_2_cities = top_cities_with_high_income.nlargest(2, 'Composite Score')

# Display the top 2 cities
print("Top 2 Cities:")
print(top_2_cities[['city', 'state_name', 'Composite Score']])


<ipython-input-9-052154776a75>:12: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

<ipython-input-9-052154776a75>:24: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

<ipython-input-9-052154776a75>:33: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

<ipython-input-9-05215477

Top 2 Cities:
         city  state_name  Composite Score
21  Las Vegas      Nevada         0.654404
12    Seattle  Washington         0.588296


<ipython-input-9-052154776a75>:45: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

<ipython-input-9-052154776a75>:53: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



if the same is done for the whole us_city_data

In [ ]:
top_cities_data = us_city_data[us_city_data['city'].isin(top_cities_with_high_income['city'])]
top_cities_data.head(5)


city city_ascii     city_alt state_id   state_name  county_fips  \
12    Seattle    Seattle          NaN       WA   Washington        53033   
21  Las Vegas  Las Vegas          NaN       NV       Nevada        32003   
22  Baltimore  Baltimore          NaN       MD     Maryland        24510   
23  St. Louis  St. Louis  Saint Louis       MO     Missouri        29510   
29   San Juan   San Juan          NaN       PR  Puerto Rico        72127   

   county_name county_fips_all county_name_all      lat  ...  poverty  \
12        King           53033            King  47.6211  ...     10.1   
21       Clark           32003           Clark  36.2333  ...     14.7   
22   Baltimore           24510       Baltimore  39.3051  ...     19.6   
23   St. Louis           29510       St. Louis  38.6359  ...     20.2   
29    San Juan           72127        San Juan  18.3985  ...      NaN   

    limited_english  commute_time  health_uninsured veteran  cbsa_fips  \
12              4.8          27.2               4.5     3.8    42660.0   
21              6.4          25.9              13.1     8.2    29820.0   
22              2.0          30.1               5.5     5.3    12580.0   
23              1.8          23.0              10.1     5.9    41180.0   
29             59.7          25.2               8.8     2.6    41980.0   

                           cbsa_name  cbsa_metro  csa_fips  \
12       Seattle-Tacoma-Bellevue, WA        True     500.0   
21  Las Vegas-Henderson-Paradise, NV        True     332.0   
22     Baltimore-Columbia-Towson, MD        True     548.0   
23                  St. Louis, MO-IL        True     476.0   
29       San Juan-Bayamón-Caguas, PR        True     490.0   

                                          csa_name  
12                              Seattle-Tacoma, WA  
21                         Las Vegas-Henderson, NV  
22  Washington-Baltimore-Arlington, DC-MD-VA-WV-PA  
23         St. Louis-St. Charles-Farmington, MO-IL  
29                            San Juan-Bayamón, PR  

[5 rows x 91 columns]

In [ ]:
# Define weights for each race category
weight_white = 0.2
weight_black = 0.2
weight_asian = 0.3
weight_native = 0.05
weight_pacific = 0.05
weight_other = 0.05
weight_multiple = 0.05
weight_hispanic = 0.1

# Combine race columns to create 'Total Demographics' with weighted sum
filtered_cities['Total Demographics'] = (
    filtered_cities['race_white'] * weight_white +
    filtered_cities['race_black'] * weight_black +
    filtered_cities['race_asian'] * weight_asian +
    filtered_cities['race_native'] * weight_native +
    filtered_cities['race_pacific'] * weight_pacific +
    filtered_cities['race_other'] * weight_other +
    filtered_cities['race_multiple'] * weight_multiple +
    filtered_cities['hispanic'] * weight_hispanic
)

# Combine economic indicators to create 'Economic Indicators'
filtered_cities['Economic Indicators'] = filtered_cities['income_household_median'] - filtered_cities['rent_median']

# Combine CBSA metro flag, commute time, and CSA information to create 'Geographic Factors'
# Define weights for each component
weight_cbsa_metro = 0.2
weight_commute_time = 0.1
weight_csa = 0.1

# Create 'Geographic Factors' column
filtered_cities['Geographic Factors'] = (
    (filtered_cities['cbsa_metro'].astype(int) * weight_cbsa_metro) +
    (1 - filtered_cities['commute_time'] / filtered_cities['commute_time'].max()) * weight_commute_time +
    (filtered_cities['csa_fips'].notnull().astype(int) * weight_csa)
)

# Drop individual race columns since they are no longer needed
filtered_cities.drop(columns=['race_white', 'race_black', 'race_asian', 'race_native', 'race_pacific', 'race_other', 'race_multiple', 'hispanic'], inplace=True)
print(filtered_cities.head())


         city city_ascii     city_alt state_id   state_name  county_fips  \
12    Seattle    Seattle          NaN       WA   Washington        53033   
21  Las Vegas  Las Vegas          NaN       NV       Nevada        32003   
22  Baltimore  Baltimore          NaN       MD     Maryland        24510   
23  St. Louis  St. Louis  Saint Louis       MO     Missouri        29510   
29   San Juan   San Juan          NaN       PR  Puerto Rico        72127   

   county_name county_fips_all county_name_all      lat  ...  \
12        King           53033            King  47.6211  ...   
21       Clark           32003           Clark  36.2333  ...   
22   Baltimore           24510       Baltimore  39.3051  ...   
23   St. Louis           29510       St. Louis  38.6359  ...   
29    San Juan           72127        San Juan  18.3985  ...   

    health_uninsured  veteran  cbsa_fips                         cbsa_name  \
12               4.5      3.8    42660.0       Seattle-Tacoma-Bellevue, WA   
21

In [ ]:
 # Normalize the composite columns
from sklearn.preprocessing import MinMaxScaler

# Initialize the MinMaxScaler
scaler = MinMaxScaler()

# Normalize the composite columns
filtered_cities[['Total Demographics', 'Economic Indicators', 'Geographic Factors', 'population', 'income_household_median']] = scaler.fit_transform(filtered_cities[['Total Demographics', 'Economic Indicators', 'Geographic Factors', 'population', 'income_household_median']])

# weights
weight_demographics = 0.1
weight_economic = 0.2
weight_geographic = 0.1
weight_population = 0.3
weight_income = 0.3

# composite score for each city
filtered_cities['Composite Score'] = (filtered_cities['Total Demographics'] * weight_demographics) + (filtered_cities['Economic Indicators'] * weight_economic) + (filtered_cities['Geographic Factors'] * weight_geographic) + (filtered_cities['population'] * weight_population) + (filtered_cities['income_household_median'] * weight_income)

# Rank the cities and select the top 2
top_2_cities = filtered_cities.nlargest(2, 'Composite Score')

# Display the top 2 cities
print("Top 2 Cities:")
print(top_2_cities[['city', 'state_name', 'Composite Score']])

Top 2 Cities:
             city  state_name  Composite Score
12        Seattle  Washington         0.687363
3466  Short Hills  New Jersey         0.666788


## Findings
Metrics used:
Population
Income
Demographics
Economic indicators
geographic factors

case 1:

If we first sort all the cities in the us_city-data based on population and income and exclude the cities which already have a NBA team in that state
Then we assign others as weights to the model we got

Top 2 Cities:

         city  state_name  Composite Score
21  Las Vegas      Nevada          0.56769

12    Seattle  Washington          0.56210


case 2:

If we directly consider all the factors as weights for the model we got

Top 2 Cities:

Top 2 Cities:

             city  state_name  Composite Score

12        Seattle  Washington         0.687363

3466  Short Hills  New Jersey         0.666788



In [ ]:
# # weights
# weights = {
#     'population': 0.2,
#     'income_household_median': 0.2,
#     'age_median': 0.15,
#     'unemployment_rate': 0.1,
#     'home_value': 0.1,
#     'rent_median': 0.1,
#     'commute_time': 0.15
# }

# # Normalization
# normalized_data = top_cities_data.copy()
# for column in normalized_data.columns:
#     if column.lower() in weights:
#         normalized_data[column] = (normalized_data[column] - normalized_data[column].min()) / (normalized_data[column].max() - normalized_data[column].min())

# # Composite score
# composite_score = (normalized_data[list(weights.keys())] * list(weights.values())).sum(axis=1)

# #ranking cities
# ranked_cities = top_cities_data.copy()
# ranked_cities['Composite Score'] = composite_score
# ranked_cities = ranked_cities.sort_values(by='Composite Score', ascending=False)

# top_two_cities = ranked_cities.head(2)
# print("Top Recommended Cities for New NBA Teams:")
# for index, city in top_two_cities.iterrows():
#     print(f"{city['city']}, {city['state_name']} (Composite Score: {city['Composite Score']})")
